### PR TITLE
[Speed up] Use passive form to select.select + Fix pipetools.uts

### PR DESCRIPTION
**Pull request by gpotter2**
Hello !
I've talked from this earlier: here's an improvment on the way of simulating select.select.

That should also fix the PyPy bug on `pipetools.uts`

- What happened before:
When using `select_objects`, an active thread would iterate through the different provided objects and call continously `check_recv`. This takes lots of memory usage.

- Now:
When using `select_objects`, the objects are all iterated once, to check if they are ready to read. If not, a thread for each object is started. Each thread will instantly block itself using Lock(), which do not consume memory. Those threads are released when the object become available, because their `call_release` function gets called, then call a callback. On top of that, a master thread waits (passively) for any callback to be fired, then returns the results.

### CODE REVIEW

**Comment by gpotter2** [comment]
@p-l- As you might have seen, some linux tests are crashing because of pipetools tests. This patch should also resolve this...

**Comment by gpotter2** [comment]
(I'm testing that it does no crash, even after many builds)

**Comment by gpotter2** [comment]
@p-l- Hey, this should fix the pipetool.uts travis/appveyor bugs.

File: scapy/pipetool.py [hunk 1]
```diff
@@ -70,12 +70,13 @@ def fileno(self):
         return self.__fdr
 
     def _read_cmd(self):
-        self.__fd_queue.pop()
-        return os.read(self.__fdr,1)
+        os.read(self.__fdr,1)
+        return self.__fd_queue.popleft()
```
Comment (gpotter2) [inline scapy/pipetool.py]: I switched those to avoid the pipetool bug

**Comment by gpotter2** [comment]
Here's a very fancy doc.

File: scapy/automaton.py [hunk 2]
```diff
@@ -19,27 +19,70 @@
...
 class SelectableObject:
+    """DEV: to implement one of those, you need to add 2 things to your object:
+    - add "check_recv" function
+    - call "self.call_release" once you are ready to be read"""
     trigger = threading.Lock()
     was_ended = False
     def check_recv(self):
         """DEV: will be called only once (at beggining) to check if the object is ready."""
-        return False
+        raise OSError("This function must be overwriten.")
```
Comment (ThomasFaivre) [inline scapy/automaton.py]: use `NotImplementedError` instead?

Comment (gpotter2) [inline scapy/automaton.py]: Hum, no idea. This just should not happen... For me `NotImplementedError` would be something that we have not developped yet, whereas this means that the API call is wrong.

File: scapy/automaton.py [hunk 3]
```diff
@@ -19,17 +19,169 @@
...
+"""
+
+class SelectableObject:
+    """DEV: to implement one of those, you need to add 2 things to your object:
+    - add "check_recv" function
+    - call "self.call_release" once you are ready to be read"""
+    trigger = threading.Lock()
+    was_ended = False
+    def check_recv(self):
+        """DEV: will be called only once (at beggining) to check if the object is ready."""
```
Comment (guedou) [inline scapy/automaton.py]: beginning

File: scapy/automaton.py [hunk 4]
```diff
@@ -19,17 +19,169 @@
...
+        raise OSError("This function must be overwriten.")
```
Comment (guedou) [inline scapy/automaton.py]: s/function/method/

File: scapy/automaton.py [hunk 5]
```diff
@@ -19,17 +19,169 @@
 from scapy.supersocket import SuperSocket
 from scapy.consts import WINDOWS
 
-class ObjectPipe:
+# In Windows, select.select is not available for custom objects. Here's the implementation of scapy to re-create this functionnality
```
Comment (guedou) [inline scapy/automaton.py]: Move this comment to the doctring.

File: scapy/automaton.py [hunk 6]
```diff
@@ -19,17 +19,169 @@
...
+            return callback(self)
+        threading.Thread(target=self._wait_non_ressources, args=(callback,)).start()
+        
+    def call_release(self, arborted=False):
+        """DEV: Must be call when the object becomes ready to read."""
+        # Relesases the lock of _wait_non_ressources
+        self.was_ended = arborted
+        try:
+            self.trigger.release()
+        except:
```
Comment (guedou) [inline scapy/automaton.py]: Could you specify the correct exception type ?

File: scapy/automaton.py [hunk 7]
```diff
@@ -19,17 +19,169 @@
...
+    - add "check_recv" function
+    - call "self.call_release" once you are ready to be read"""
+    trigger = threading.Lock()
+    was_ended = False
+    def check_recv(self):
+        """DEV: will be called only once (at beggining) to check if the object is ready."""
+        raise OSError("This function must be overwriten.")
+
+    def _wait_non_ressources(self, callback):
+        # This get started as a thread, and waits for the data lock to be freed
```
Comment (guedou) [inline scapy/automaton.py]: Can you use doctrings instead of comments ? (everywhere)

File: scapy/automaton.py [hunk 8]
```diff
@@ -19,17 +19,169 @@
...
+        self.inputs = list(inputs)
+        self.remain = remain
+        self.available_lock = threading.Lock()
+        self.available_lock.acquire()
+        self._ended = False
+
+    def process(self):
+        # Entry point of SelectableSelector
+        if WINDOWS:
+            if not self.remain:
```
Comment (guedou) [inline scapy/automaton.py]: Why not replacing the two loops with:
```
for i in self.inputs:
  if not isinstance(i, SelectableObject):
    warning("Unknown ignored object type: " + type(i))
  elif not self.remain and if i.check_recv():
    self.results.append(i)
  else:
    i.wait_return(self._exit_door)
```

**Comment by guedou** [comment]
@p-l- what's your call ?

**Pull request merged**